lib/core/src/Cardano/Wallet/Primitive/Types.hs

```diff
@@ -999,9 +1005,27 @@ newtype SlotNo = SlotNo { unSlotNo :: Word32 }
     deriving stock (Show, Read, Eq, Ord, Generic)
     deriving newtype (Num, Buildable, NFData, Enum)
 
-newtype EpochNo = EpochNo { unEpochNo :: Word64 }
+newtype EpochNo = EpochNo { unEpochNo :: Word31 }
     deriving stock (Show, Read, Eq, Ord, Generic)
-    deriving newtype (Num, Buildable, NFData, Enum)
+    deriving newtype (Num, Bounded, Enum)
+
+instance Buildable EpochNo where
+    build (EpochNo e) = build $ fromIntegral @Word31 @Word32 e
+
+instance NFData EpochNo where
+    rnf (EpochNo !_) = ()
+
+-- | Convert the specified value into an 'EpochNo', or fail if the value is
+--   too large.
+unsafeEpochNo :: HasCallStack => Word32 -> EpochNo
+unsafeEpochNo epochNo
+    | epochNo > maxEpochNo =
+        error "unsafeEpochNo: epoch number out of bounds."
+    | otherwise =
+        EpochNo $ fromIntegral epochNo
+  where
+    maxEpochNo :: Word32
+    maxEpochNo = fromIntegral @Word31 $ unEpochNo maxBound
```

Comment:
My first instinct was to locate this functton within the `Unsafe` module. However, doing so introduces a circular import.